Add the ability to list an application's routing table

settermjd opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

The change, in short, adds a new Mezzio command: routes:list. When invoked, it will print the application's routing table akin to what [Symfony](https://symfony.com/doc/7.0/routing.html#debugging-routes) and [Laravel](https://laravel.com/docs/11.x/routing#listing-your-routes) provide.

The command's output can be:

- Printed in two formats: tabular and JSON
- Sorted based on the route's name or path
- Filtered based on whether a route has a name, path, middleware class, or supports a given method

**Note:** This was originally created in #27, but that PR got quite messy. So, my intention is to add the code in this new PR without the baggage of the original.

settermjd committed:
Add the ability to list an application's routing table

This was originally created in #27, but that PR got quite messy. So, my
intention is to add the code in a new PR with the baggage of the
original.

The change, in short, adds a new Mezzio command: routes:list. When
invoked, it will print the application's routing table akin to what
Symfony and Laravel provide.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
I'm having a few issues getting Psalm to pass, and would really appreciate help addressing these issues. I've corrected as many as I can, but am not experience enough with Psalm to correct the remaining handful.

settermjd committed:
Update README to conform with Markdown-Lint rules

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Minor corrections to address Psalm rules

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

alexmerlin commented on test/Routes/RoutesFilterTest.php:
```diff
@@ -0,0 +1,227 @@
+        );
+    }
+
+    /**
```
```suggestion
    /**
     * @param array<string, mixed> $filterOptions
```

alexmerlin commented on test/Routes/RoutesFilterTest.php:
```diff
@@ -0,0 +1,227 @@
+        );
+    }
+
+    public function validFilterDataProvider(): array
```
```suggestion
    /**
     * @psalm-return array<array-key, array{0: int, 1: array<string, mixed>}>
     */
    public function validFilterDataProvider(): array
```

alexmerlin commented on test/Routes/ListRoutesCommandTest.php:
```diff
@@ -0,0 +1,496 @@
+     */
+    private function reflectExecuteMethod(): ReflectionMethod
+    {
+        $r = new ReflectionMethod($this->command, 'execute');
```
No need to call `$r->setAccessible(true);` so you can replace method body with:
```php
return new ReflectionMethod($this->command, 'execute');
```

settermjd replied:
I did that locally and will commit the change shortly.

alexmerlin replied:
> No need to call `$r->setAccessible(true);` so you can replace method body with:
> 
> ```
> return new ReflectionMethod($this->command, 'execute');
> ```

This is the last missing piece.

alexmerlin commented on test/Routes/ListRoutesCommandTest.php:
```diff
@@ -0,0 +1,496 @@
+     * @dataProvider sortRoutingTableDataProvider
+     * @throws ReflectionException
+     */
+    public function testCanSortResults(string $sortOrder, string $expectedOutput): void
```
```suggestion
    public function testCanSortResults(string $sortOrder): void
```
Parameter `$expectedOutput` is not used.

alexmerlin commented on test/Routes/ListRoutesCommandTest.php:
```diff
@@ -0,0 +1,496 @@
+    /**
+     * @dataProvider filterRoutingTableDataProvider
+     */
+    public function testCanFilterRoutingTable(array $filterOptions, string $expectedOutput): void
```
```suggestion
    public function testCanFilterRoutingTable(array $filterOptions): void
```
Parameter `$expectedOutput` is not used.

alexmerlin commented on test/Routes/ListRoutesCommandTest.php:
```diff
@@ -0,0 +1,496 @@
+    }
+
+    /**
+     * @return array<array-key, array<array-key, mixed>>
```
```suggestion
     * @psalm-return array<array-key, array<array<string, string>>>
```
From what I see, the JSON is not used anywhere, you can remove it - in that case my suggestion is valid - else, let me know and I can provide a different `psalm-return` that covers that string as well.

alexmerlin commented on test/Routes/ListRoutesCommandTest.php:
```diff
@@ -0,0 +1,496 @@
+        return [
+            [
+                ['middleware' => 'ExpressMiddleware'],
+                '[{"name":"home","path":"\/","methods":"GET","middleware":"MezzioTest\\\\Tooling\\\\Routes\\\\Middleware\\\\ExpressMiddleware"}]',
```
```suggestion
```

alexmerlin commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,187 @@
+ * those being the route's name, path, middleware, or supported method(s).
+
+ * @template TKey
+ * @template-covariant TValue
```
```suggestion
 * @template-covariant TValue
 * @template TIterator as Iterator<TKey, TValue>
```

alexmerlin commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,187 @@
+
+ * @template TKey
+ * @template-covariant TValue
+ * @template-extends FilterIterator<TKey, TValue>
```
```suggestion
 * @template-extends FilterIterator<TKey, TValue, TIterator>
```

alexmerlin commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,187 @@
+
+use ArrayIterator;
+use Exception;
+use FilterIterator;
```
```suggestion
use FilterIterator;
use Iterator;
```

alexmerlin reviewed (commented):
I debugged/tested locally until everything was green.
With these changes all checks should pass.

settermjd commented:
Thanks for the assist, @alexmerlin

settermjd committed:
Update test/Routes/RoutesFilterTest.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update test/Routes/RoutesFilterTest.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update test/Routes/ListRoutesCommandTest.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update test/Routes/ListRoutesCommandTest.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update src/Routes/Filter/RoutesFilter.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update src/Routes/Filter/RoutesFilter.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update src/Routes/Filter/RoutesFilter.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Merge branch 'mezzio:2.10.x' into add-command-to-print-routing-table

settermjd committed:
Update test/Routes/ListRoutesCommandTest.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update test/Routes/ListRoutesCommandTest.php

Co-authored-by: Alex Karajos <alex.merlin.1985@gmail.com>
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Remove unnecessary call to ReflectionMethod::setAccessible

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Remove unnecessary terminal output

On running it, it doesn't seem to make much sense to tell the user what
they can already see (and what they asked for). Given that, the terminal
output stating the routing table's format type is being removed in this
change.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Update the routes:list command's documentation

This is a small update to provide better examples of what the user can
expect from running the routes:list command.

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
All feedback has been integrated, and the documentation has been updated (in the README) to better show the user what to expect from the command.

settermjd commented:
Once merged, I'll submit a PR to https://docs.mezzio.dev/mezzio/v3/reference/cli-tooling/ to reflect the changes.

alexmerlin commented:
> All feedback has been integrated, and the documentation has been updated (in the README) to better show the user what to expect from the command.

This one is missing:

```php
return new ReflectionMethod($this->command, 'execute');
```

it fixes [this issue](https://github.com/mezzio/mezzio-tooling/pull/52/files#diff-7bf86cf849563d0803c9f3c588845808f9317a3b26e2f714362232ece6ddde74R61) reported by psalm in `test/Routes/ListRoutesCommandTest.php`

alexmerlin commented:
Also, a missing new line at the end of README.md
https://github.com/mezzio/mezzio-tooling/actions/runs/11365400646/job/31613519235?pr=52#step:3:355

settermjd committed:
Remove unnecessary local variable in reflectExecuteMethod

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Add newline at the end of the README

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
> Also, a missing new line at the end of README.md https://github.com/mezzio/mezzio-tooling/actions/runs/11365400646/job/31613519235?pr=52#step:3:355

Done.

settermjd commented:
> > All feedback has been integrated, and the documentation has been updated (in the README) to better show the user what to expect from the command.
> 
> This one is missing:
> 
> ```
> return new ReflectionMethod($this->command, 'execute');
> ```
> 
> it fixes [this issue](https://github.com/mezzio/mezzio-tooling/pull/52/files#diff-7bf86cf849563d0803c9f3c588845808f9317a3b26e2f714362232ece6ddde74R61) reported by psalm in `test/Routes/ListRoutesCommandTest.php`

Done.

alexmerlin commented:
If you want all checks to be green, now there's an extra new line at the end of README.md that needs to be removed.
https://github.com/mezzio/mezzio-tooling/actions/runs/11397057942/job/31711903304?pr=52#step:3:355

settermjd commented:
> If you want all checks to be green, now there's an extra new line at the end of README.md that needs to be removed. https://github.com/mezzio/mezzio-tooling/actions/runs/11397057942/job/31711903304?pr=52#step:3:355

Oddly, it was complaining about one not being there, previously. I'll remove it.

settermjd committed:
Remove unnecessary blank line at the end of the README

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd committed:
Correct code style break in README

Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

settermjd commented:
Fixed the final issues. Thanks for your patience, @alexmerlin.

gsteel commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * Match the route against a regular expression based on the field in $matchType.
+     *
+     * $matchType can be either "path" or "name".
```
```suggestion
     * @param 'name'|'path' $routeAttribute
```

gsteel commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,189 @@
+     *
+     * $matchType can be either "path" or "name".
+     */
+    public function matchesByRegex(Route $route, string $routeAttribute): bool
```
```suggestion
    private function matchesByRegex(Route $route, string $routeAttribute): bool
```

gsteel commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * Match if the current route supports the method(s) supplied.
+     */
+    public function matchesByMethod(Route $route): bool
```
```suggestion
    private function matchesByMethod(Route $route): bool
```

gsteel commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,189 @@
+     * the class' name. The intent is to perform checks from the least to the
+     * most computationally expensive, to avoid excessive overhead.
+     */
+    public function matchesByMiddleware(Route $route): bool
```
```suggestion
    private function matchesByMiddleware(Route $route): bool
```

gsteel commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * @param ArrayIterator<int, Route> $routes
+     */
+    public function __construct(ArrayIterator $routes, array $filterOptions = [])
```
Why is $filterOptions an array when you have the otherwise unused `RouteFilterOptions` class?

settermjd replied:
Good question. I honestly do not know.

gsteel commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,240 @@
+    /** @var null|string Cannot be defined explicitly due to parent class */
+    public static $defaultName = 'mezzio:routes:list';
+
+    public function __construct(
+        ContainerInterface $container,
+        ConfigLoaderInterface $configLoader
+    ) {
```
Why inject the container when you only appear to be fetching `RouteCollector`? Would it not be better to inject only the RouteCollectorInterface here?

gsteel commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,240 @@
+use function strtolower;
+use function usort;
+
+class ListRoutesCommand extends Command
```
Shouldn't Symfony commands have the `#[AsCommand('whatever')]` attribute these days? Pretty sure this available in v6 and required in v7

gsteel commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,240 @@
+        return $result;
+    }
+
+    public function getRows(bool $requireNames = false): array
```
This method should probably be private. Also, would it be better to pass in the `RoutesFilter` as an argument? Then you wouldn't be accruing state in instance properties

gsteel commented on src/Routes/ListRoutesCommand.php:
```diff
@@ -0,0 +1,240 @@
+        return $rows;
+    }
+
+    public function getSortOrder(InputInterface $input): string
```
private?

gsteel commented on src/Routes/ListRoutesCommandFactory.php:
```diff
@@ -0,0 +1,26 @@
+        /** @var MiddlewareFactory $factory */
+        $factory = $container->get(MiddlewareFactory::class);
+
+        $configLoader = new RoutesFileConfigLoader('config/routes.php', $application, $factory, $container);
```
Would it not be better to simply call `$container->get(ConfigLoaderInterface::class)`, create a factory for the 'Default' config loader and then alias the default config loader to the interface in container factory config. This way, users would only need to override the alias with a custom loader implementation. Creating a concrete loader here means that users would need to write a new factory for the command if they are loading routes in a different way. I for one don't have a `config/routes.php` file in any of my apps.

settermjd replied:
That makes perfect sense to me. It seems that I made some assumptions that I didn't check, or just didn't think about some things.

gsteel commented on test/Routes/ListRoutesCommandFactoryTest.php:
```diff
@@ -0,0 +1,33 @@
+use PHPUnit\Framework\TestCase;
+use Psr\Container\ContainerInterface;
+
+class ListRoutesCommandFactoryTest extends TestCase
```
Symfony ships with a command tester implementation that can be used to assert output of the command - this could be used to at least verify the different output formats?

gsteel commented on src/Routes/Filter/RouteFilterOptions.php:
```diff
@@ -0,0 +1,88 @@
+        string $middleware = '',
+        string $name = '',
+        string $path = '',
+        $methods = []
```
```suggestion
        array|string $methods = []
```

gsteel commented on src/Routes/Filter/RouteFilterOptions.php:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * @param string|array<array-key,string> $methods
+     */
+    public function __construct(
```
Generally speaking, I would prefer to see no default options here and instead unions of non-empty-string|null perhaps - that way it's explicit that null means the option is unset and should be ignored, and otherwise, you have a non-empty-string to work with.

settermjd replied:
I don't quite follow. Do you mean something like:

```
public function __construct(
    private ?string $middleware = null,
    private ?string $name = null,
    private ?string $path = null,
    $methods = null
)
```

gsteel commented on src/Routes/Filter/RoutesFilter.php:
```diff
@@ -0,0 +1,189 @@
+                    sprintf('/%s/', $matchesMiddleware),
+                    $middlewareClass
+                );
+        } catch (Exception $e) {
```
Should probably be `Throwable`? and I think you can drop the `$e` from PHP 8.1 onwards

gsteel reviewed (commented):
Just an initial pass with a few questions :)

gsteel commented:
Oops - didn't spot this had been merged…

froschdesign commented:
@settermjd 
All hints from @gsteel must be addressed in a separate pull request.

settermjd commented:
> @settermjd All hints from @gsteel must be addressed in a separate pull request.

I figured.

settermjd commented:
> Oops - didn't spot this had been merged…

All good. I only saw your comments just now.

settermjd commented:
@gsteel, first up, thank you for all of your feedback. I really appreciate it and look forward to learning a lot through implementing changes in response to it.

settermjd commented:
@gsteel / @froschdesign the PR is almost done.
